### DISCUSSION TITLE
Using TensorRT file (end2end.engine) outside MMDeploySDK

### DISCUSSION THREAD

**Discussion by k-ishi8** [open-mmlab/mmdeploy#1247] [Q&A] [answered]
I would like to use the end2end.engine(MMPose TensorRT) obtained by MMDeploy with Triton Inference Server.
I am aware that end2end.engine only includes inference processing, so I think I need to implement the preprocessing myself, is that correct?
Also, I am outputting pipeline.json to understand the contents of the preprocessing, but I have a question about the value of Normalize.

I used [this](https://github.com/open-mmlab/mmdeploy/blob/master/docs/en/04-supported-codebases/mmpose.md) as a reference to get end2end.engine and pipeline.json.
In config, I specified topdown_heatmap/coco/hrnet_w48_coco_256x192.py.
In topdown_heatmap/coco/hrnet_w48_coco_256x192.py, type='NormalizeTensor' mean=[0.485, 0.456, 0.406],std=[0.229, 0.224, 0.225]. In pipeline.json, "type": "Normalize","mean": [123.675,116.28,103.53],"std": [58.395,57.120000000000005,57.375] is set.
Is there a difference between 'NormalizeTensor' and 'Normalize'?

**Comment by lzhangzz** [comment, marked as answer]
You can find the definition of `NormalizeTensor` [here](https://github.com/open-mmlab/mmpose/blob/8c58a18b138b0aa12ed3ce341954707f9236c036/mmpose/datasets/pipelines/shared_transform.py#L48).

It seems that `NormalizeTensor` is working on `torch.Tensor` in CHW format, while `Normalize` is working on HWC `numpy.Array`

**Reply by k-ishi8** [reply]
@lzhangzz Thank you for your response.
If I use end2end.engine, should I convert with HWC or CHW?
Since what is listed in `Pipeline.json` is `Normalize`, should I convert it to HWC?

**Reply by lvhan028** [reply]
As far as I know, `pipeline.json` usually includes `Normalize` followed by `ImageToTensor`.
`Normalize` request the input being in `HWC` format.
After `Normalize`, `Image2Tensor` will transpose the tensor from `HWC` to `CHW`.

**Reply by k-ishi8** [reply]
Thank you for your response.
I was able to get the preprocessed data using [Compose](https://mmpose.readthedocs.io/en/latest/api.html#mmpose.datasets.pipelines.shared_transform.Compose) and test_pipeline.
